service/history/shardController.go

```diff
@@ -322,24 +322,31 @@ func (c *shardController) acquireShards() {
 	sw := c.metricsClient.StartTimer(metrics.HistoryShardControllerScope, metrics.AcquireShardsLatency)
 	defer sw.Stop()
 
-AcquireLoop:
-	for shardID := 0; shardID < c.config.NumberOfShards; shardID++ {
-		info, err := c.hServiceResolver.Lookup(string(shardID))
-		if err != nil {
-			c.logger.Error("Error looking up host for shardID", tag.Error(err), tag.OperationFailed, tag.ShardID(shardID))
-			continue AcquireLoop
-		}
+	concurrency := c.config.AcquireShardConcurrency()
+	sem := make(chan bool, concurrency)
 
-		if info.Identity() == c.host.Identity() {
-			_, err1 := c.getEngineForShard(shardID)
-			if err1 != nil {
-				c.metricsClient.IncCounter(metrics.HistoryShardControllerScope, metrics.GetEngineForShardErrorCounter)
-				c.logger.Error("Unable to create history shard engine", tag.Error(err1), tag.OperationFailed, tag.ShardID(shardID))
-				continue AcquireLoop
+	for shardID := 0; shardID < c.config.NumberOfShards; shardID++ {
+		sem <- true
+		go func(shardID int) {
+			defer func() { <-sem }()
+			info, err := c.hServiceResolver.Lookup(string(shardID))
+			if err != nil {
+				c.logger.Error("Error looking up host for shardID", tag.Error(err), tag.OperationFailed, tag.ShardID(shardID))
+				return
 			}
-		} else {
-			c.removeEngineForShard(shardID)
-		}
+			if info.Identity() == c.host.Identity() {
+				_, err1 := c.getEngineForShard(shardID)
+				if err1 != nil {
+					c.metricsClient.IncCounter(metrics.HistoryShardControllerScope, metrics.GetEngineForShardErrorCounter)
+					c.logger.Error("Unable to create history shard engine", tag.Error(err1), tag.OperationFailed, tag.ShardID(shardID))
+				}
+			} else {
+				c.removeEngineForShard(shardID)
```

Comment:
what do you think about the idea of prioritizing loads over removes ? i.e. as you are iterating through the shardIDs, if a shard is no longer owned, don't call removeEngineForShard() immediately, but store the shardID in a slice. Once all the loading is complete, then you can go over the slice and remove the shards (if the shard still needs to be removed). 

Doesn't have to be in this same diff, but if you like the idea, let's do this in a follow up diff.

Reply:
Makes sense, we can run this experiment too.